### PR TITLE
Initial drop for intent based recommendation

### PR DESCRIPTION
**Pull request by miteshvp**
This is still R&D. It needs to be more robust and remove some hardcodedness.

### CODE REVIEW

File: f8a_worker/workers/recommender.py [hunk 1]
```diff
@@ -66,15 +66,15 @@ def str_value_cleaner(safe_string):
     def execute_gremlin_dsl(self, payload):
         """Execute the gremlin query and return the response."""
         try:
-            response = get_session_retry().post(self._bayesian_graph_url, data=json.dumps(payload))
+            response = requests.post(self._bayesian_graph_url, data=json.dumps(payload))
```
Comment (samuzzal-choudhury) [inline f8a_worker/workers/recommender.py]: Has to be replaced with get_session_retry().

Comment (miteshvp) [inline f8a_worker/workers/recommender.py]: Done. Nice catch

File: f8a_worker/dispatcher/flows/stackApiGraphV2Flow.yml [hunk 2]
```diff
@@ -0,0 +1,14 @@
+---
+  flow-definitions:
+    - name: 'stackApiGraphV2Flow'
```
Comment (fridex) [inline f8a_worker/dispatcher/flows/stackApiGraphV2Flow.yml]: Is it needed to keep v1 and v2 version in git? Can we directly change the existing version?

File: f8a_worker/dispatcher/flows/stackApiGraphV2Flow.yml [hunk 3]
```diff
@@ -0,0 +1,14 @@
...
+      queue: '{DEPLOYMENT_PREFIX}_api_stackApiGraphFlow_v0'
+      edges:
+        - from:
+          to:
+            - 'GraphAggregatorTask'
+            - 'ManifestKeeperTask'
```
Comment (fridex) [inline f8a_worker/dispatcher/flows/stackApiGraphV2Flow.yml]: Plase place this to separate edges:
```yaml
  - from:
    to: 'GraphAggregatorTask'
  - from:
    to: 'ManifestKeeperTask'
```

File: f8a_worker/dispatcher/flows/stackApiGraphV2Flow.yml [hunk 4]
```diff
@@ -0,0 +1,14 @@
...
+        - from:
+            - 'GraphAggregatorTask'
+          to:
+            - 'recommendation_v2'
+            - 'stack_aggregator_v2'
```
Comment (fridex) [inline f8a_worker/dispatcher/flows/stackApiGraphV2Flow.yml]: Separate edges as stated above:

```yaml
 - from: 'GraphAggregatorTask'
   to: 'recommendation_v2'
 - from: 'GraphAggregatorTask'
   to: 'stack_aggregator_v2'
```

File: f8a_worker/graphutils.py [hunk 5]
```diff
@@ -181,3 +181,73 @@ def aggregate_stack_data(stack, manifest_file, ecosystem):
...
+        pkg_dict = {
+            'ecosystem': epv['ver']['pecosystem'][0],
+            'name': name,
+            'version': epv['ver']['version'][0],
+            'licenses': epv['ver'].get('licenses', []),
+            'sentiment': {"overall_score": 0.65, 'latest_comment': 'N/A'},
+            'latest_version': epv['pkg']['latest_version'][0],
+            'security': [],
+            'osio_user_count': 0}
+        github_dict = {
```
Comment (fridex) [inline f8a_worker/graphutils.py]: Why there are hard-coded values?

File: f8a_worker/workers/recommender.py [hunk 6]
```diff
@@ -478,3 +498,110 @@ def _get_stack_values(self, similar_stacks_list):
...
         return similarity_list
+
+
+class RecommendationV2Task(BaseTask):
+    _analysis_name = 'recommendation_v2'
+    description = 'Get Recommendation'
+
+    def call_pgm(self, payload):
+        """Calls the PGM model with the normalized manifest information to get the relevant packages"""
+        PGM_REST_URL = "http://kronos-kattappa-0239.dev.rdu2c.fabric8.io/api/v1/kronos_score"
```
Comment (fridex) [inline f8a_worker/workers/recommender.py]: Configurable vie env variables?

File: f8a_worker/workers/recommender.py [hunk 7]
```diff
@@ -478,3 +498,110 @@ def _get_stack_values(self, similar_stacks_list):
...
+                else:
+                    json_response = response.json()
+                    return json_response
+            else:
+                self.log.debug('Payload information is not passed in the call, Quitting! PGM\'s call')
+        except:
+            self.log.error("Failed retrieving PGM data.")
+            return None
+
+    def execute(self, arguments=None):
```
Comment (fridex) [inline f8a_worker/workers/recommender.py]: Missing strict assert on parent tasks:

```python
self._strict_assert('GraphAggregatorTask' in self.parent.keys())
```

File: f8a_worker/workers/stackaggregator_v2.py [hunk 8]
```diff
@@ -0,0 +1,162 @@
...
+
+import os
+import json
+import requests
+
+from f8a_worker.base import BaseTask
+from f8a_worker.graphutils import GREMLIN_SERVER_URL_REST
+from f8a_worker.utils import get_session_retry
+
+def extract_component_details(component):
```
Comment (fridex) [inline f8a_worker/workers/stackaggregator_v2.py]: Why this and aggregate_stack_data` functions are not methods?

File: f8a_worker/workers/stackaggregator_v2.py [hunk 9]
```diff
@@ -0,0 +1,162 @@
...
+                "total_licenses": len(stack_distinct_licenses),
+                "distinct_licenses": list(stack_distinct_licenses),
+                "stack_license_conflict": False,
+                "dependencies": dependencies
+            }
+    }
+    return data
+
+class StackAggregatorV2Task(BaseTask):
+    description = 'Aggregates stack data from components'
```
Comment (fridex) [inline f8a_worker/workers/stackaggregator_v2.py]: `description` field is not needed anymore.

File: f8a_worker/workers/stackaggregator_v2.py [hunk 10]
```diff
@@ -0,0 +1,162 @@
...
+                    self.log.error("Failed retrieving dependency data.")
+                    continue
+            except:
+                self.log.error("Error retrieving dependency data.")
+                continue
+
+        return {"result": result}
+
+    def execute(self, arguments=None):
+        finished = []
```
Comment (fridex) [inline f8a_worker/workers/stackaggregator_v2.py]: Missing strict assert on `GraphAggregatorTask`.

File: f8a_worker/workers/stackaggregator_v2.py [hunk 11]
```diff
@@ -0,0 +1,162 @@
...
+
+        for result in aggregated['result']:
+            resolved = result['details'][0]['_resolved']
+            ecosystem = result['details'][0]['ecosystem']
+            manifest = result['details'][0]['manifest_file']
+
+            finished = self._get_dependency_data(resolved,ecosystem)
+            if finished != None:
+                stack_data = aggregate_stack_data(finished, manifest, ecosystem.lower(), resolved)
+                print (json.dumps(stack_data))
```
Comment (fridex) [inline f8a_worker/workers/stackaggregator_v2.py]: Space after `print` shouldn't be there.

Comment (fridex) [inline f8a_worker/workers/stackaggregator_v2.py]: Event print shouldn't be present.

**Comment by miteshvp** [comment]
@fridex - all your comments are valid, and am aware of it. I have already mentioned in my comment earlier, that it needs to be more robust and have to remove hardcodedness as well. This PR was to demonstrate our Kronos Scoring engine in addition to our traditional reference stacks recommendation. Frankly speaking this should go as WIP, if not for weekly demonstrations on staging environment.

**Comment by samuzzal-choudhury** [comment]
@fridex there is reason behind being fast, isn't it? The reason is a demo to be shown in 50 mins on staging. @krishnapaparaju will keep me honest on this.

I know that things are not in proper shape. But, as I mentioned, this entire setup is experimental which can be plugged out with minimal effect to the existing codebase.

Having said that, your comments are always valuable and we will incorporate those when we go for the actual prod work.

**Comment by fridex** [comment]
> @fridex - all your comments are valid, and am aware of it. I have already mentioned in my comment earlier, that it needs to be more robust and have to remove hardcodedness as well. This PR was to demonstrate our Kronos Scoring engine in addition to our traditional reference stacks recommendation. Frankly speaking this should go as WIP, if not for weekly demonstrations on staging environment.

>@fridex there is reason behind being fast, isn't it? The reason is a demo to be shown in 50 mins on staging. @krishnapaparaju will keep me honest on this.

>I know that things are not in proper shape. But, as I mentioned, this entire setup is experimental which can be plugged out with minimal effect to the existing codebase.

>Having said that, your comments are always valuable and we will incorporate those when we go for the actual prod work.

@miteshvp @samuzzal-choudhury note that we have dev cluster that was dedicated for demos and experiments. Also note that everything that is in staging goes to production sooner or later.

**Comment by msrb** [comment]
@miteshvp @samuzzal-choudhury if you guys think this is still in WIP shape, then please bear in mind that things from staging environment will eventually end up in prod. We don't do releases and therefore we should try to keep our master branch stable at all times. Stable master branch ensures that people can push fixes to stage/prod quickly, if necessary.

If you think that this could affect stability of our production environment, I'd suggest reverting the commit after the weekly demo.

**Comment by miteshvp** [comment]
> If you think that this could affect stability of our production environment, I'd suggest reverting the commit after the weekly demo.

This will not affect the stability of the current production since we have made sure to not touch any existing working parts. We are just adding so that two recommendations (intent based and reference stacks based) can co-exist.

>if you guys think this is still in WIP shape, then please bear in mind that things from staging environment will eventually end up in prod.

I too raised the same concerns so I will have to pass this for @krishnapaparaju to answer this.

**Comment by krishnapaparaju** [comment]
All are valid comments.. Is ths Dev. cluster upgrade done ?

**Comment by fridex** [comment]
> Is ths Dev. cluster upgrade done?

Yes, it's running the same OpenShift version as in staging/prod.

**Pull request merged**